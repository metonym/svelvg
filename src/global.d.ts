declare module "estree-walker" {
  import type { Element } from "svelte/types/compiler/interfaces";

  export function walk(
    ast: Ast,
    options: {
      enter: (node: Element, parentNode: Element) => void;
    },
  ): void;
}
