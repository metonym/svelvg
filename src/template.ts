import { parse, walk } from "svelte/compiler";

interface Position {
  start: number;
  end: number;
}

interface NodeElement extends Position {
  type: "Element";
  name: string;
  children: Position[];
  attributes: {
    name: string;
    value: { raw: any }[];
  }[];
}

export const templateSvelte = (
  source: string,
  filename: string,
  { classes }: { classes: string[] }
) => {
  const ast = parse(source, { filename });

  let svg_attributes = "";
  let svg_children = "";

  classes.forEach((name) => (svg_attributes += ` class:${name}={true}`));

  walk(ast.html, {
    enter(node: NodeElement) {
      if (node.type === "Element" && node.name === "svg") {
        node.children.forEach((child) => {
          svg_children += source.slice(child.start, child.end);
        });

        node.attributes.forEach(({ name, value }) => {
          if (name === "class") {
            value[0].raw.split(" ").forEach((name: string) => {
              svg_attributes += ` class:${name}={true}`;
            });
          } else {
            svg_attributes += ` ${name}="${value[0].raw}"`;
          }
        });
      }
    },
  });

  return `<svg${svg_attributes} {...$$restProps}><slot />${svg_children.trim()}</svg>`;
};

export const templateTs = (
  moduleName: string
) => `/// <reference types="svelte" />
import type { SvelteComponentTyped } from "svelte";

export interface ${moduleName}Props
  extends svelte.JSX.SVGAttributes<SVGSVGElement> {}

export default class ${moduleName} extends SvelteComponentTyped<
  ${moduleName}Props,
  {},
  { default: {} }
> {}
`;
