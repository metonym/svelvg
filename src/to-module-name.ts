export const toModuleName = (name: string) =>
  name
    .replace(/\-+/g, " ")
    .replace(/\.svg/g, "")
    .split(" ")
    .map((_, i) => {
      let first_char = _.slice(0, 1);

      if (i === 0 && first_char.match(/[0-9]|!/)) {
        first_char = "_" + first_char;
      }

      return first_char.toUpperCase() + _.slice(1);
    })
    .join("");
