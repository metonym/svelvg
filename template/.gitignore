.DS_Store
dist
lib
node_modules