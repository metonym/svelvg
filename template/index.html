<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  </head>
  <body>
    <script type="module">
      import IconLibrary from "svelvg/lib/IconLibrary.svelte";
      import * as icons from "./lib";
      import pkg from "./package.json";

      const app = new IconLibrary({
        target: document.body,
        props: { pkg, icons },
      });
    </script>
  </body>
</html>
