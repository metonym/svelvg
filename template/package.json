{
  "name": "svelvg",
  "version": "0.1.0",
  "license": "MIT",
  "description": "lib-name SVG icons as Svelte components",
  "author": "Eric Liu (https://github.com/metonym)",
  "svelte": "./lib/index.js",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "vite build --base=/svelvg/",
    "prepack": "svelvg glob=bootstrap-icons/icons iconIndex",
    "test": "bun test"
  },
  "devDependencies": {
    "@sveltejs/vite-plugin-svelte": "^3.1.2",
    "bootstrap-icons": "^1.11.3",
    "svelte": "^4.2.19",
    "svelvg": "^0.12.2",
    "vite": "^5.4.2"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/metonym/npm-name.git"
  },
  "homepage": "https://github.com/metonym/npm-name",
  "bugs": "https://github.com/metonym/npm-name/issues",
  "keywords": [
    "svelte",
    "svelte component",
    "icon",
    "svg"
  ],
  "files": [
    "lib"
  ]
}
