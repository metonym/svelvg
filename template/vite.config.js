import { svelte } from "@sveltejs/vite-plugin-svelte";
import { defineConfig } from "vite";
import { ghPages } from "vite-plugin-gh-pages";

export default defineConfig({
  // replace with "npm-name"
  base: "/svelvg/",
  plugins: [
    svelte({
      compilerOptions: { dev: false, immutable: true },
    }),
    ghPages(),
  ],
});
