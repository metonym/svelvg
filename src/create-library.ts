import tg from "tiny-glob";
import fs from "fs/promises";
import { existsSync } from "fs";
import path from "path";
import { templateSvelte, templateTs } from "./template";
import { toModuleName } from "./to-module-name";

interface CreateLibraryOptions {
  /** @default "lib" */
  outDir: string;

  /** @default false */
  iconIndex: boolean | string;

  /**
   * Callback to add a list of classes to the SVG element
   * provided the original filename and module name
   * @example
   * filename: "alarm-fill"
   * moduleName: "AlarmFill"
   */
  appendClassNames: (
    filename: string,
    moduleName: string
  ) => undefined | string[];

  /**
   * Override the default module name
   */
  toModuleName: (params: {
    path: path.ParsedPath;
    moduleName: string;
  }) => string;
}

type ModuleNames = string[];

export const createIndexFile = (moduleNames: ModuleNames) =>
  moduleNames
    .map(
      (moduleName) =>
        `export { default as ${moduleName} } from "./${moduleName}.svelte";\n`
    )
    .join("");

export const createIconIndex = (moduleNames: ModuleNames) => `# Icon Index

> ${moduleNames.length} total icons

## Icons

${moduleNames.map((moduleName) => `- ${moduleName}\n`).join("")}`;

export async function createLibrary(
  glob: string,
  options: Partial<CreateLibraryOptions>
) {
  const outDir = options?.outDir ?? "lib";

  let iconIndex: boolean | string = false;

  if (typeof options?.iconIndex === "string") {
    iconIndex = options.iconIndex;
  } else if (options?.iconIndex === true) {
    iconIndex = "ICON_INDEX.md";
  }

  let libPath = glob;

  if (!/^node_modules/.test(libPath)) {
    libPath = path.join("node_modules", glob);
  }

  if (!/\*.svg$/.test(libPath)) {
    libPath = path.join(libPath, "*.svg");
  }

  const libDir = path.parse(libPath).dir;

  try {
    await fs.stat(libDir);
  } catch (error) {
    console.error(`[createLibrary] path does not exist: ${libDir}`);
    process.exit(1);
  }

  try {
    const dir = path.join(process.cwd(), outDir);

    if (existsSync(dir)) {
      await fs.rm(dir, { recursive: true });
    }
    await fs.mkdir(dir);

    const files = await tg(libPath.replace(/\*.svg$/, "**/*.svg"));
    const moduleNames: ModuleNames = [];

    for (const filename of files) {
      const source = await fs.readFile(filename, "utf-8");
      const parsedPath = path.parse(filename);

      let moduleName = toModuleName(parsedPath.name);

      if (options?.toModuleName) {
        moduleName = options.toModuleName({ path: parsedPath, moduleName });
      }

      const classes =
        options?.appendClassNames?.call(null, parsedPath.name, moduleName) ??
        [];

      fs.writeFile(
        path.join(dir, `${moduleName}.svelte`),
        templateSvelte(source, filename, { classes })
      );
      fs.writeFile(
        path.join(dir, `${moduleName}.svelte.d.ts`),
        templateTs(moduleName)
      );

      moduleNames.push(moduleName);
    }

    const uniqueModuleNames = [...new Set(moduleNames)];
    const index = createIndexFile(uniqueModuleNames);

    fs.writeFile(path.join(dir, "index.js"), index);
    fs.writeFile(path.join(dir, "index.d.ts"), index);

    console.log(
      `⚡ Converted ${uniqueModuleNames.length} icons from "${glob}" to Svelte components in "${outDir}"`
    );

    if (iconIndex) {
      fs.writeFile(
        path.join(process.cwd(), iconIndex),
        createIconIndex(uniqueModuleNames)
      );

      console.log(`✏️ Wrote icon index to "${iconIndex}"`);
    }
  } catch (error) {
    console.error(error);
  }
}
