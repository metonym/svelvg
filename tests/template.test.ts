import { test, expect } from "vitest";
import { templateSvelte, templateTs } from "../src/template";

test("templateSvelte", () => {
  [
    {
      name: "Element only",
      source: `<svg></svg>`,
      filename: undefined,
      classes: [],
    },
    {
      name: "With attributes and children",
      source: `<svg id="id" class="class"><title>Title</title></svg>`,
      filename: undefined,
      classes: [],
    },
    {
      name: "With classes",
      source: `<svg id="id" class="class"><title>Title</title></svg>`,
      filename: undefined,
      classes: ["my-class"],
    },
  ].forEach(({ source, filename, classes }) => {
    expect(templateSvelte(source, filename, { classes })).toMatchSnapshot();
  });
});

test("templateTs", () => {
  expect(templateTs("IconName")).toMatchSnapshot();
});
