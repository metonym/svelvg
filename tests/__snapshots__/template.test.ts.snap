// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`templateSvelte 1`] = `"<svg {...$$restProps}><slot /></svg>"`;

exports[`templateSvelte 2`] = `"<svg id="id" class:class={true} {...$$restProps}><slot /><title>Title</title></svg>"`;

exports[`templateSvelte 3`] = `"<svg class:my-class={true} id="id" class:class={true} {...$$restProps}><slot /><title>Title</title></svg>"`;

exports[`templateTs 1`] = `
"export { IconName as default } from "./";
"
`;
